### DISCUSSION TITLE
How to access kernel global variables

### DISCUSSION THREAD

**Discussion by iomartin** [bpftrace/bpftrace#3327] [Q&A]
Hi,

Is it possible to access kernel global variables from my bpftrace program? My searches seem to indicate that yes (ksym?), but I can't find an example on how to do it.

In particular, I want to access the vector_irq variable from [common_interrupt](https://elixir.bootlin.com/linux/v5.15/source/arch/x86/kernel/irq.c#L248):

```C
DEFINE_IDTENTRY_IRQ(common_interrupt)
{
	struct pt_regs *old_regs = set_irq_regs(regs);
	struct irq_desc *desc;

	/* entry code tells RCU that we're not quiescent.  Check it. */
	RCU_LOCKDEP_WARN(!rcu_is_watching(), "IRQ failed to wake up RCU");

	desc = __this_cpu_read(vector_irq[vector]);
	(...)
}
```

Thanks!

**Comment by jordalgo** [comment]
Hi @iomartin,

Thanks for your question. `ksym` is used to symbolize a kernel address not access a global kernel variable. AFAIK bpftrace cannot access global kernel variables and TBH, I'm not sure how'd you accomplish this in a raw BPF program. To access kernel variables you'd need a pointer to them passed as part of some probe arguments or context.

**Comment by ajor** [comment]
What you're looking for is the `kaddr()` function - it's the inverse of `ksym()`. It translates a symbol name into an address: https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#kaddr (please forgive what looks like a truncated documentation entry).

Once you've got the address all you need to do is dereference it to read the variable's data. The [`loads.bt`](https://github.com/bpftrace/bpftrace/blob/44019baafb9fb78d123e72be3c3c28ed361a1db3/tools/loads.bt#L31-L32) tool uses it to read the value of the `avenrun` global variable. You should be able to do something similar for `vector_irq`. Feel free to cast the pointer too, e.g. `$vector_irq = *(struct irq_desc**)kaddr("vector_irq");` (I may have got the type wrong there, but you hopefully get the idea)

**Reply by iomartin** [reply]
Thanks @ajor, this is exactly what I was looking for!

However I'm getting NULL with `kaddr("vector_irq")`, probably because these are per-cpu varibles. Do you know how I could access them?

My goal is to do `__this_cpu_read(vector_irq[vector])` for different `vector` values

Thanks!

**Reply by ajor** [reply]
Just to make sure: this script is returning a non-null address, right? `BEGIN { printf("%p\n", kaddr("vector_irq")) }`
(it should match the output of `grep vector_irq /proc/kallsyms`)

To actually read from your per-cpu variable, I think it's just a matter of finding the correct offsets.

`vector_irq`'s definition is:
```
typedef struct irq_desc* vector_irq_t[NR_VECTORS];
DECLARE_PER_CPU(vector_irq_t, vector_irq);
```
So the `vector_irq` variable will have been declared with a type something like:
`__percpu struct irq_desc* vector_irq[256]`

Which is equivalent to:
`__percpu struct irq_desc** vector_irq`

And we start off with a pointer to that with the address from `kaddr`, i.e.:
`__percpu struct irq_desc*** vector_irq`

I don't know what offsets you'll need to read from to get into the correct per-cpu memory (that'll take a lot of kernel macro reading), but hopefully this is a start:
```
BEGIN
{
  $vector_irq = (struct irq_desc**)*kaddr("vector_irq");
  $vector = 12; // pick some index into the array
  $myvector = *($vector_irq + $vector); // pointer arithmetic for array indexing
}
```

**Reply by iomartin** [reply]
Thanks @ajor! I don't know why my `kaddr("vector_irq")` was coming up as NULL before, but it seems to be getting the right address now (it matches /proc/kallsyms).

I'll go through the macros to get the per-cpu offset.

**Reply by danobi** [reply]
There are bpf helpers to access percpu variables. It would be great if bpftrace supported them

https://github.com/torvalds/linux/blob/786c8248dbd33a5a7a07f7c6e55a7bfc68d2ca48/include/uapi/linux/bpf.h#L4917-L4944

**Reply by danobi** [reply]
#1837. I seemed to have believed at the time that some BTF magic is necessary. Perhaps that is true. But regardless, since we use libbpf to load now, it should work
